docs/internal/op-sem.md

```diff
@@ -0,0 +1,438 @@
+    - if we do change it, then it looks like `paused~>heap` instead of `paused~>iso` if we read ourself from our parent region
+    - if we `undef` our storageloc in order to call `region`, this is fine
+  - imm
+    - easy, it's in the object header
+
+## Types
+
+Assignable fields are `store T`. Embed fields and methods are `T`.
+
+```
+Γ       ∈ TypeEnv     ::= x: T | Γ₁, Γ₂ | α
+T       ∈ Type        ::= α
+                        | iso | mut | imm | paused | stack
+                        | (T₁ | T₂)
+                        | (T₁ & T₂)
+                        | (T₁, T₂, ...)
+                        | {f: T}
+                        | store T
+                        | ClassID
+                        | Γ₁ ⊢ T₁→T₂ ⊣ Γ₂
+```
+
+## Regions
+
+```ts
+imm: no region
+  <- ∅
+  -> { mut, stack }
+iso: disjoint region
+  <- ∅
+  -> { mut, stack }
+mut: top region
+  <- { iso, imm, mut }
+  -> { mut, stack }
+stack: all regions
+  <- { iso, imm, mut, paused, stack }
+  -> { stack }
+paused: subset of regions not including top
+  <- ∅
+  -> { stack }
+```
+
+## Definitions
+
+```
+x, y, z ∈ Id
+τ       ∈ TypeId
+P       ∈ Program     ::= (Id → Function) × (TypeId → Type)
+          Type        ::= TypeId* × (Id → Member)
+m       ∈ Member      ::= Function | Id
+ϕ       ∈ Frame       ::= Region* × (Id → Value) × Id* × Expression*
+σ       ∈ State       ::= Frame*
+                        × (ObjectId → Object)
+                        × (StorageLoc → Value)
+                        × (Region → Strategy)
+                        × Bool
+v       ∈ Value       ::= ObjectId
+                        | StorageLoc
+                        | Function
+                        | Bool
+                        | Undefined
+ι       ∈ ObjectId
+f       ∈ StorageLoc  ::= ObjectId × Id
+          Bool        ::= true | false
+λ         Function    ::= Id* × Expression*
+undef   ∈ Undefined
+ω       ∈ Object      ::= Region* × TypeId
+ρ       ∈ Region
+Σ       ∈ Strategy    ::= GC | RC | Arena
+e       ∈ Expression  ::= x = var
+                        | x = dup y
+                        | x = load y
+                        | x = store y z
+                        | x = lookup y z
+                        | x = typetest x τ
+                        | x = new τ
+                        | x = stack τ
+                        | x* = call y(y*)
+                        | x* = region y(z, z*)
+                        | x* = create Σ y(z*)
+                        | tailcall x(x*)
+                        | branch x y(y*) z(z*)
+                        | return x*
+                        | error
+                        | x = catch
+                        | acquire x
+                        | release x
+                        | release v
+                        | fulfill x
+```
+
+## Shape
+
+```ts
+σ₁, e₁* → σ₂, e₂*
+```
+
+## Simplifying Notation
+
+```ts
+// Program operations.
+P.functions = P↓₁
+P.types = P↓₂
+P.types(τᵩ) = ((), [x↦x])
```

Comment:
I don't understand what's going on here, sorry, especially with `ω <: τ = τ ∈ P.types(ω.type)↓₁` below.  `τ ∈ ()` looks like something that should fail to type check?  (`ω(x) = P.types(ω.type)↓₂(x)` at least type checks, but I'll confess I'm not sure what's going on there either.)